### PR TITLE
Updating dependencies, rework of client provider

### PR DESCRIPTION
**Pull request by ivelten**
With this Pull Request, I am updating all dependencies to their last possible versions, and migrating the build script to FAKE 5. There is also a rewrite of the client provider to solve some design and dependencies issues.

- [x] Design changes to the provider (query string based)
- [x] Removing dependencies on JSON.NET
- [x] Add XML docs to mapped GraphQL Types
- [x] Make an object for other types of included information (other than `data` and `errors`)
- [x] Mark GraphQL deprecated fields as deprecated on provided types
- [x] Make asynchronous run methods
- [x] Map variables to method parameters
- [x] Update documentation with provider usage
- [x] Check client provider package installation
- [x] Make it possible to provide custom HTTP headers
- [x] Make it possible to get introspection via GET method
- [x] Unit testing of new provider

### CODE REVIEW

File: tests/FSharp.Data.GraphQL.Tests/FSharp.Data.GraphQL.Tests.fsproj [hunk 1]
```diff
@@ -51,5 +50,8 @@
     <ProjectReference Include="..\..\src\FSharp.Data.GraphQL.Server\FSharp.Data.GraphQL.Server.fsproj" />
     <ProjectReference Include="..\..\src\FSharp.Data.GraphQL.Server.Middlewares\FSharp.Data.GraphQL.Server.Middlewares.fsproj" />
   </ItemGroup>
+  <ItemGroup Condition="$(OS) == Unix">
```
Comment (ivelten) [inline tests/FSharp.Data.GraphQL.Tests/FSharp.Data.GraphQL.Tests.fsproj:57]: For some reason, running xUnit tests with `dotnet test` in Unix/OSX machines gives an error in Remoting (reference: https://github.com/xunit/xunit/issues/1357).

Adding a `xunit.runner.json` file with an `"appDomain": "denied"` entry seems to solve this problem as recomended on the issue link.

**Comment by purkhusid** [comment]
Is this PR still being worked on?

**Comment by ivelten** [comment]
> Is this PR still being worked on?

Yes, @purkhusid. Initially we were going to just update some dependencies and make the actual client provider be cross environment compatible by trying to adjust its dependencies on Newtonsoft.Json package.

But we thought a bit and decided that the actual client provider design is not as useful as we would like to, and decided to rewrite it completely using a new design, and having no package dependencies.

In a nutshell, you  can se how the new design is becoming by looking at the files [test1.fsx](https://github.com/fsprojects/FSharp.Data.GraphQL/blob/71e2498aaff89b46311546efe1713406c212bab0/samples/client-provider/test1.fsx) and [test2.fsx](https://github.com/fsprojects/FSharp.Data.GraphQL/blob/71e2498aaff89b46311546efe1713406c212bab0/samples/client-provider/test2.fsx) on the client provider samples folder.

To help you giving track about where is this PR at the moment, I made a checklist on the PR description.

**Comment by purkhusid** [comment]
Awesome, really looking forward to testing it out. It should work with Fable since the query TP is erasing, right?

**Comment by ivelten** [comment]
> Awesome, really looking forward to testing it out. It should work with Fable since the query TP is erasing, right?

That's the purpose, @purkhusid. We made it erasing to be able to work with Fable in the future. But we won't test it out until the new version comes out, though.

**Comment by NatElkins** [comment]
I just watched a talk on this library from the recent FSharpX conference, really interesting.  My company is using F# with GraphQL in production right now, but we're just using some wrappers around https://github.com/graphql-dotnet/graphql-client .  I was taking a look at the test scripts that show how the new client provider will be used.  I was wondering, does it support arguments in the query?  The example I looked at had an input parameter hard-coded.

**Comment by johnberzy-bazinga** [comment]
@NatElkins Great to hear you watched @Horusiath talk. Check this fsx out that uses parameters.

https://github.com/fsprojects/FSharp.Data.GraphQL/blob/305efb947fe2d562987b51de877652c191cf9f7c/samples/client-provider/test4.fsx

**Comment by NatElkins** [comment]
@johnberzy-bazinga Very cool!

Has any thought been given to ways to compose queries?  For example, suppose you had a `product` entity and wanted to request the same properties on it in multiple queries.  Right now in my code base we have something like this:

```fsharp
    let skuBuilder (baseQuery:Builder) =
        baseQuery
            .add(Sku.ID)
            .add(Sku.BRAND)
            .add(Sku.HAS_RESOLVED_BRAND)
            .add(Sku.NAME)
            .add(Sku.ASIN)
            .add(Sku.GTIN)
            .add(Sku.UPC)
            .add(Sku.PRICE)
            .add(Sku.SOURCE)
            .add(Sku.PRODUCT_URL)
            .add(Sku.SOURCE_IMAGE)
            .add(Sku.IS_AVAILABLE)
            .add(Sku.UPDATED)
            .add(Sku.ATTRIBUTES)
            .add(Sku.RECOMMENDATION_COUNT)
            .add(Sku.RECOMMENDED_OR_AGREED_USER_COUNT)
            .add(Builder(Sku.CATEGORIES)
                    .add(Category.ID)
                    .add(Category.NAME)
                    .add(Category.PATH_STRING))

    let! (mskuDetails:JsonValue[]) =
        skuBuilder
        |> RecommendedSku.Query.get(mskuIds)
        |> Async.map (Result.toOption >> Option.getOrElse [||])
```

to build up queries.  The nice thing about this is the `skuBuilder` is reusable from other GraphQL queries we want to make.

I am keeping a close eye on this.  Please let me know if I can do anything to help!

**Comment by johnberzy-bazinga** [comment]
@NatElkins that's an interesting idea. I'm not sure how we'd accomplish that with stringified TPs. We'd probably need to wait for something like  https://github.com/fsharp/fslang-design/issues/125 and have some sort of mechanism to compose multiple TP instances?

**Pull request merged**